libs/git_wrapper/Git_wrapper.ml

```diff
@@ -631,6 +626,94 @@ let cat_file_blob ?cwd sha =
   | Error (`Msg s) ->
       Error s
 
+let obj_type_of_string = function
+  | "commit" -> Some Commit
+  | "blob" -> Some Blob
+  | "tree" -> Some Tree
+  | "tag" -> Some Tag
+  | _ -> None
+
+let batch_cat_file_blob ?cwd blob_shas =
+  let cmd =
+    Bos.Cmd.(
+      v "git"
+      %% of_list
+           (cd cwd
+           @ [
+               "cat-file";
+               (* Print object information and contents for each object
+                  provided on stdin. *)
+               "--batch";
+               (* Since we write to a file, ensure output is buffered. If we
+                  change to read from stdio, we probably should _not_ buffer so
+                  we can stream output. *)
+               "--buffer";
+               (* Null terminate the header line. Requires git >2.42.0. *)
+             ]))
+  in
+  (* Each object is formatted as
+
+         <oid> SP <type> SP <size> NUL
+         <contents> LF
+
+     See also
+     <https://git-scm.com/docs/git-cat-file#:~:text=For%20example%2C%20%2D%2Dbatch%20without%20a%20custom%20format%20would%20produce%3A>
+
+     This gets the next output out of the channel, and then leaves the rest.
+     This is so we can generate a lazy Seq.t so we don't have to read all of
+     the objects into memory if we are going to write them back out to
+     tempfiles sequentially anyway.
+  *)
+  let get_next_obj chan =
+    (* Read the header line *)
+    match In_channel.input_line chan with
+    | None ->
+        (* Remember to close the channel when we're done.
+           Note that this means the user needs to consume the entire generated
+           stream currently in order to avoid a resource leak, but I think this
+           is fine.
+        *)
+        close_in chan;
+        None
+    | Some metadata ->
+        Some
+          ( (match String.split_on_char ' ' metadata with
+            | [ sha; kind; size ] ->
+                let ( let* ) = Result.bind in
+                let* kind =
+                  obj_type_of_string kind
+                  |> Option.to_result
+                       ~none:
+                         (spf "invalid object type %s for object %s" kind sha)
+                in
+                let* size =
+                  int_of_string_opt size
+                  |> Option.to_result
+                       ~none:
+                         (spf "invalid object size %s for object %s" size sha)
+                in
+                (* Now use the size from the header line to read the whole
+                   contents all at once *)
+                let contents = really_input_string chan size in
+                (* discard trailing newline *)
+                ignore (input_char chan);
+                let obj = { kind; sha; extra = { contents } } in
+                Ok obj
+            | _ -> Error ("invalid git object: " ^ metadata)),
+            chan )
+  in
+  let output = UCommon.new_temp_file "git-batch-cat-files" ".log" |> Fpath.v in
```

Comment:
(or use a variant of `with_file`)

Reply:
We cannot use a `with_file` style function (well maybe, since the channel should still work if the deletion is just `unlink`, but I think it less clear) since the Seq relies on this file existing. And `new_temp_file` registers this for deletion, I believe.